src/RouteAggregator.js

```diff
@@ -0,0 +1,132 @@
+import invariant from 'invariant';
+import Relay from 'react-relay';
+
+import getParamsForRoute from './getParamsForRoute';
+
+export default class RouteAggregator {
+  constructor() {
+    this.uniqueQueryNames = new WeakMap();
+    this.lastQueryIndex = 0;
+
+    this.route = null;
+
+    this.data = {};
+    this.failure = null;
+  }
+
+  updateRoute(routerProps) {
+    const {branch, params, location} = routerProps;
+
+    const relayRoute = {
+      queries: {},
+      params: {},
+    };
+    const fragmentSpecs = {};
+
+    branch.forEach(route => {
+      const {queries} = route;
+      if (!queries) {
+        return;
+      }
+
+      const {component} = route;
+
+      // In principle not all container component routes have to specify
+      // queries, because some of them might somehow receive fragments from
+      // their parents, but it would definitely be wrong to specify queries
+      // for a component that isn't a container.
+      invariant(
+        Relay.isContainer(component),
+        'relay-nested-routes: Route with queries specifies component `%s` ' +
+        'that is not a Relay container.',
+        component.displayName || component.name
+      );
+
+      const routeParams = getParamsForRoute({route, branch, params, location});
+      Object.assign(relayRoute.params, routeParams);
+
+      Object.keys(queries).forEach(queryName => {
+        const query = queries[queryName];
+        const uniqueQueryName = this.getUniqueQueryName(query, queryName);
+
+        relayRoute.queries[uniqueQueryName] =
+          () => query(component, routeParams);
+        fragmentSpecs[uniqueQueryName] = {component, queryName};
+      });
+    });
+
+    relayRoute.name =
+      ['aggregated', ...Object.keys(relayRoute.queries)].join('-');
+
+    // RootContainer uses referential equality to check for route change, so
+    // replace the route object entirely.
+    this.route = relayRoute;
+    this.fragmentSpecs = fragmentSpecs;
+  }
+
+  getUniqueQueryName(query, queryName) {
+    let uniqueQueryName = this.uniqueQueryNames.get(query);
+    if (uniqueQueryName === undefined) {
```

Comment:
I actually can't - the airbnb ESLint config errors on `==` anywhere, including for comparisons to `null`. I figured doing `=== undefined` is clearer here because it shouldn't be possible for `null` values to end up in the map at all.